huaweicloud/services/dws/resource_huaweicloud_dws_workload_plan_stage.go

```diff
@@ -0,0 +1,357 @@
+		product = "dws"
+	)
+
+	client, err := cfg.NewServiceClient(product, region)
+	if err != nil {
+		return diag.Errorf("error creating DWS client: %s", err)
+	}
+
+	createPath := client.Endpoint + httpUrl
+	createPath = strings.ReplaceAll(createPath, "{project_id}", client.ProjectID)
+	createPath = strings.ReplaceAll(createPath, "{cluster_id}", d.Get("cluster_id").(string))
+	createPath = strings.ReplaceAll(createPath, "{plan_id}", d.Get("plan_id").(string))
+	createOpt := golangsdk.RequestOpts{
+		KeepResponseBody: true,
+		JSONBody:         utils.RemoveNil(buildCreateWorkloadStageBodyParams(d)),
+	}
+
+	resp, err := client.Request("POST", createPath, &createOpt)
+	if err != nil {
+		return diag.Errorf("error creating DWS workload plan stage: %s", err)
+	}
+	respBody, err := utils.FlattenResponse(resp)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	respErrCode := utils.PathSearch("error_code", respBody, "")
+	if respErrCode != "" {
+		return diag.Errorf("error creating DWS workload plan stage: %s", utils.PathSearch("workload_res_str", respBody, nil))
+	}
+	return resourceWorkLoadPlanStageRead(ctx, d, meta)
+}
+
+func buildCreateWorkloadStageBodyParams(d *schema.ResourceData) map[string]interface{} {
+	return map[string]interface{}{
+		"workload_plan_stage": map[string]interface{}{
+			"stage_name": d.Get("name"),
+			"month":      d.Get("month"),
+			"day":        d.Get("day"),
+			"start_time": d.Get("start_time"),
+			"end_time":   d.Get("end_time"),
+			"queue_list": buildCreateQueueListBodyParam(d),
+		},
+	}
+}
+
+func buildCreateQueueListBodyParam(d *schema.ResourceData) []map[string]interface{} {
+	queueList := d.Get("queue_list").([]interface{})
+	if len(queueList) == 0 {
+		return nil
+	}
+	params := make([]map[string]interface{}, 0, len(queueList))
+	for _, queue := range queueList {
+		queueObject := queue.(map[string]interface{})
+		param := map[string]interface{}{
+			"queue_name":      queueObject["name"],
+			"queue_resources": buildQueueConfigParam(queueObject["configuration"]),
+		}
+		params = append(params, param)
+	}
+	return params
+}
+
+func buildQueueConfigParam(d interface{}) []map[string]interface{} {
+	rawParams := d.(*schema.Set)
+	if rawParams.Len() == 0 {
+		return nil
+	}
+	params := make([]map[string]interface{}, 0, rawParams.Len())
+	for _, rawParam := range rawParams.List() {
+		raw := rawParam.(map[string]interface{})
+		param := map[string]interface{}{
+			"resource_name":        raw["resource_name"],
+			"resource_value":       raw["resource_value"],
+			"value_unit":           raw["value_unit"],
+			"resource_description": raw["resource_description"],
+		}
+		params = append(params, param)
+	}
+
+	return params
+}
+
+func resourceWorkLoadPlanStageRead(_ context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	var (
+		cfg     = meta.(*config.Config)
+		region  = cfg.GetRegion(d)
+		httpUrl = "v2/{project_id}/clusters/{cluster_id}/workload/plans/{plan_id}"
```

Comment:
why not use the API for querying stage details

Reply:
the plan details API contains all stage information and does not require further queries

Reply:
done